src/libtrx/game/items.c

```diff
@@ -113,6 +114,117 @@ bool Item_GetAnimChange(ITEM *const item, const ANIM *const anim)
     return false;
 }
 
+void Item_Translate(
+    ITEM *const item, const int32_t x, const int32_t y, const int32_t z)
+{
+    const int32_t c = Math_Cos(item->rot.y);
+    const int32_t s = Math_Sin(item->rot.y);
+    item->pos.x += ((c * x + s * z) >> W2V_SHIFT);
+    item->pos.y += y;
+    item->pos.z += ((c * z - s * x) >> W2V_SHIFT);
+}
+
+void Item_Animate(ITEM *const item)
+{
+    item->hit_status = 0;
+    item->touch_bits = 0;
+    item->frame_num++;
+
+    const ANIM *anim = Item_GetAnim(item);
+    if (anim->num_changes > 0 && Item_GetAnimChange(item, anim)) {
+        anim = Item_GetAnim(item);
+        item->current_anim_state = anim->current_anim_state;
+
+        if (item->required_anim_state == anim->current_anim_state) {
+            item->required_anim_state = 0;
+        }
+    }
+
+    if (item->frame_num > anim->frame_end) {
+        for (int32_t i = 0; i < anim->num_commands; i++) {
+            const ANIM_COMMAND *const command = &anim->commands[i];
+            switch (command->type) {
+            case AC_MOVE_ORIGIN: {
+                const XYZ_16 *const pos = (XYZ_16 *)command->data;
+                Item_Translate(item, pos->x, pos->y, pos->z);
+                break;
+            }
+
+            case AC_JUMP_VELOCITY: {
+                const ANIM_COMMAND_VELOCITY_DATA *const data =
+                    (ANIM_COMMAND_VELOCITY_DATA *)command->data;
+                item->fall_speed = data->fall_speed;
+                item->speed = data->speed;
+                item->gravity = true;
+                break;
+            }
+
+            case AC_DEACTIVATE:
+                item->status = IS_DEACTIVATED;
+                break;
+
+            default:
+                break;
+            }
+        }
+
+        item->anim_num = anim->jump_anim_num;
+        item->frame_num = anim->jump_frame_num;
+        anim = Item_GetAnim(item);
+
+#if TR_VERSION == 1
+        item->current_anim_state = anim->current_anim_state;
+        item->goal_anim_state = item->current_anim_state;
+#else
+        if (item->current_anim_state != anim->current_anim_state) {
+            item->current_anim_state = anim->current_anim_state;
+            item->goal_anim_state = anim->current_anim_state;
+        }
+#endif
```

Comment:
How significant is this difference?

Reply:
I did wonder about this, and apologies, I forgot to touch on it. I think there must be dependencies later after `Item_Animate` where we don't want to lose `goal_anim_state` on the final frame. I imagine we can just do the same for TR1, but we can perhaps tackle it on its own and test the items more thoroughly under that change.

Reply:
Thanks. I agree, let's not do it here by any means.